internal/api/impl/auth/oauth.go

```diff
@@ -282,71 +300,198 @@ func (e *oAuthEndpoint) codeExchangeHandler(ctx echo.Context) error {
 	}
 
-	uInfo, err := e.buildUserInfo(uInfoBody)
+	_, err = e.performUserSync(uInfo, ctx.SetCookie)
 	if err != nil {
-		e.logWithError(err).Error("Failed to collect user infos.")
 		return apiinterface.InternalError
 	}
 
-	// Save the user in database
-	entity, err := e.svc.syncUser(uInfo)
+	return ctx.Redirect(302, "/")
+}
+
+// deviceCodeHandler is the http handler on Perses side that will trigger the "Device Authorization"
+// flow to the oauth 2.0 provider.
+// It will return the provider's DeviceAuthResponse as is, containing some information for the
+// user to login in a browser.
+// Then the client will be responsible to poll the /{slug_id}/token Perses endpoint to generate
+// a proper Perses session.
+// TODO: Use device code config if provided
+func (e *oAuthEndpoint) deviceCodeHandler(ctx echo.Context) error {
+	resp, err := e.conf.DeviceAuth(ctx.Request().Context())
 	if err != nil {
-		e.logWithError(err).Error("Failed to sync user in database.")
 		return apiinterface.InternalError
 	}
+	return ctx.JSON(200, resp)
+}
+
+// tokenHandler is the http handler on Perses side that will generate a proper Perses session.
+// It is used only in case of device code flow and client credentials flow.
+func (e *oAuthEndpoint) tokenHandler(ctx echo.Context) error {
+	var reqBody tokenRequestBody
+	if err := ctx.Bind(&reqBody); err != nil {
+		e.logWithError(err).Error("Invalid request body")
+		return echo.NewHTTPError(http.StatusBadRequest, "Invalid request body")
+	}
+
+	var accessToken *oauth2.Token
+	var err error
+	switch reqBody.GrantType {
+	case GrantTypeDeviceCode:
+		accessToken, err = e.retrieveDeviceAccessToken(ctx.Request().Context(), reqBody.DeviceCode)
+		if err != nil || !accessToken.Valid() {
+			if err == nil {
+				// APIs like Github might return an invalid token without error
+				err = errors.New("invalid token received")
+			}
+			// (We log a warning as the failure means most of the time that the user didn't authorize the app yet)
+			e.logWithError(err).Warn("Failed to exchange device code for token")
+			return echo.NewHTTPError(http.StatusInternalServerError, "Failed to exchange device code for token")
+		}
+	case GrantTypeClientCredentials:
+		accessToken, err = e.retrieveClientCredentialsToken(ctx.Request().Context(), reqBody.ClientID, reqBody.ClientSecret)
+		if err != nil || !accessToken.Valid() {
+			if err == nil {
+				// APIs like Github might return an invalid token without error
+				err = errors.New("invalid token received")
+			}
+			e.logWithError(err).Error("Failed to exchange client credentials for token")
+			return echo.NewHTTPError(http.StatusInternalServerError, "Failed to exchange client credentials for token")
+		}
+	default:
+		return echo.NewHTTPError(http.StatusBadRequest, "Unsupported grant type")
+	}
 
-	username := entity.GetMetadata().GetName()
-	_, err = e.tokenManagement.accessToken(username, ctx.SetCookie)
+	uInfo, err := e.requestUserInfo(ctx, accessToken)
+	if err != nil {
```

Comment:
We'd probably need to create a more fine-grained error management to differentiate the errors that needs to be considered as expected (to be tried again) or fatal (a next shot will not be better)

Reply:
I materialized it into a comment. 
```go
                        // Note that for sake of simplicity, it doesn't respect the Oauth 2.0 RFC: (https://www.rfc-editor.org/rfc/rfc8628#section-3.5)
			// For example:
			// - we should differentiate the errors that means "try again later" from the others. (error: "authorization_pending")
			// - we should propagate the message possibly saying to slow down the request (error: "slow_down")
```

Not sure if we do the effort of comply fully with OAuth 2.0 specs on that. If yes maybe a larger work would be necessary to comply in all the auth endpoints error response... which is quite a lot of work oO

Reply:
As discussed, I propose to move the target of this PR to a feature branch ``feature/extend-oauth`` and I will make a separate PR to make fine-grain ``/token`` error management. Following the OAuth 2.0 RFC. 
I will also finalize the client credentials flow in that branch, through another PR.